File: runtime/src/terraclaw_runtime/skills/library/explore.py
"""Explore skill — systematically explore unknown territory."""

from __future__ import annotations

import asyncio
from typing import Any

from terraclaw_runtime.bridge.client import BridgeClient
from terraclaw_runtime.bridge.message import Action, Observation
from terraclaw_runtime.skills.registry import SkillResult


class ExploreSkill:
    name = "explore"
    description = "Explore the surrounding area. Moves in an expanding spiral pattern, reporting interesting discoveries."

    async def check_preconditions(self, obs: Observation, params: dict[str, Any]) -> tuple[bool, str]:
        return True, ""

    async def execute(self, bridge: BridgeClient, obs: Observation, params: dict[str, Any]) -> SkillResult:
        direction = params.get("direction", "right")
        steps = params.get("steps", 5)
        step_distance = params.get("step_distance_ms", 2000)

        discoveries: list[str] = []

        for i in range(steps):
            # Move in the exploration direction
            await bridge.send_action(Action(
                action_type="move",
                params={
                    "direction": direction,
                    "duration_ms": step_distance,
                    "auto_jump": True,
                },
                timeout_ms=step_distance + 3000,
            ))

            await asyncio.sleep(step_distance / 1000.0 + 0.5)

            # Observe what we found
            fresh = await bridge.receive_observation(timeout=1.0)
            if fresh:
                interesting = fresh.spatial_window.get("tiles", {}).get("interesting", [])
                for tile in interesting:
                    ttype = tile.get("type", "")
                    if ttype not in ("door", "work_bench", "pressure_plate"):
                        discoveries.append(ttype)

                # Check for enemies
                hostiles = [n for n in fresh.entities.get("npcs", []) if n.get("is_hostile")]
                if hostiles:
                    discoveries.append(f"enemies:{len(hostiles)}")

        return SkillResult(
            success=True,
            message=f"Explored {steps} steps {direction}. Found: {', '.join(discoveries) if discoveries else 'nothing notable'}",
            data={"discoveries": discoveries, "steps": steps},
        )

    async def cancel(self) -> None:
        pass

File: runtime/config/skills/explore.lua
-- explore.lua
-- Explore by flying in a direction, observing interesting features.
return {
    name = "explore",
    description = "Fly in a direction and observe surroundings",

    execute = function(ctx, params)
        local direction = params.direction or "right"
        local steps = params.steps or 5
        local step_size = params.step_size or 320 -- 20 tiles per step

        local dx = 0
        if direction == "right" then dx = 1
        elseif direction == "left" then dx = -1 end

        local results = {}
        for i = 1, steps do
            local move_result = ctx:send_action("move_to", {
                x = params.start_x + dx * i * step_size,
                y = params.start_y or 0,
                speed = 8.0,
                arrival_radius = 32.0,
            })
            ctx:wait(200)
            results[i] = move_result
        end

        return { success = true, message = "Explored " .. direction, steps = steps }
    end,
}
